CUDA 11 Test: cuDNN TestConvolutionBackwardFilter

### Failing Tests

```
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardFilter_param_159_{auto_tune=True, deterministic=False, dilate=1, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='always'}::test_call FAILED [ 33%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardFilter_param_160_{auto_tune=True, deterministic=False, dilate=1, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='auto'}::test_call FAILED [ 33%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardFilter_param_161_{auto_tune=True, deterministic=False, dilate=1, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='never'}::test_call FAILED [ 33%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardFilter_param_231_{auto_tune=True, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='always'}::test_call FAILED [ 36%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardFilter_param_232_{auto_tune=True, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='auto'}::test_call FAILED [ 36%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardFilter_param_233_{auto_tune=True, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='never'}::test_call FAILED [ 36%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardFilter_param_447_{auto_tune=False, deterministic=False, dilate=1, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='always'}::test_call FAILED [ 45%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardFilter_param_448_{auto_tune=False, deterministic=False, dilate=1, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='auto'}::test_call FAILED [ 45%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardFilter_param_449_{auto_tune=False, deterministic=False, dilate=1, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='never'}::test_call FAILED [ 45%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardFilter_param_519_{auto_tune=False, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='always'}::test_call FAILED [ 48%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardFilter_param_520_{auto_tune=False, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='auto'}::test_call FAILED [ 48%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardFilter_param_521_{auto_tune=False, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='never'}::test_call FAILED [ 48%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_231_{auto_tune=True, bias=True, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='always'}::test_call FAILED [ 60%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_232_{auto_tune=True, bias=True, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='auto'}::test_call FAILED [ 60%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_233_{auto_tune=True, bias=True, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='never'}::test_call FAILED [ 60%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_237_{auto_tune=True, bias=True, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='always'}::test_call FAILED [ 60%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_238_{auto_tune=True, bias=True, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='auto'}::test_call FAILED [ 60%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_239_{auto_tune=True, bias=True, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='never'}::test_call FAILED [ 60%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_519_{auto_tune=True, bias=False, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='always'}::test_call FAILED [ 71%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_520_{auto_tune=True, bias=False, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='auto'}::test_call FAILED [ 71%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_521_{auto_tune=True, bias=False, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='never'}::test_call FAILED [ 71%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_525_{auto_tune=True, bias=False, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='always'}::test_call FAILED [ 72%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_526_{auto_tune=True, bias=False, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='auto'}::test_call FAILED [ 72%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_527_{auto_tune=True, bias=False, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='never'}::test_call FAILED [ 72%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_807_{auto_tune=False, bias=True, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='always'}::test_call FAILED [ 83%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_808_{auto_tune=False, bias=True, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='auto'}::test_call FAILED [ 83%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_809_{auto_tune=False, bias=True, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='never'}::test_call FAILED [ 83%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_813_{auto_tune=False, bias=True, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='always'}::test_call FAILED [ 84%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_814_{auto_tune=False, bias=True, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='auto'}::test_call FAILED [ 84%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_815_{auto_tune=False, bias=True, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='never'}::test_call FAILED [ 84%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_1095_{auto_tune=False, bias=False, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='always'}::test_call FAILED [ 95%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_1096_{auto_tune=False, bias=False, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='auto'}::test_call FAILED [ 95%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_1097_{auto_tune=False, bias=False, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=0, ndim=3, tensor_core='never'}::test_call FAILED [ 95%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_1101_{auto_tune=False, bias=False, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='always'}::test_call FAILE  [ 95%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_1102_{auto_tune=False, bias=False, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='auto'}::test_call FAILED [ 95%]
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_1103_{auto_tune=False, bias=False, deterministic=False, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='never'}::test_call FAILED [ 96%]
```

### Enviornment

```
CuPy Version          : 8.0.0b5
CUDA Root             : /usr/local/cuda
CUDA Build Version    : 11000
CUDA Driver Version   : 11000
CUDA Runtime Version  : 11000
cuBLAS Version        : 11100
cuFFT Version         : 10200
cuRAND Version        : 10201
cuSOLVER Version      : (10, 5, 0)
cuSPARSE Version      : 11100
NVRTC Version         : (11, 0)
Thrust Version        : 100909
CUB Build Version     : 100909
cuDNN Build Version   : 8002
cuDNN Version         : 8002
NCCL Build Version    : None
NCCL Runtime Version  : None
cuTENSOR Version      : None
```

```
    -------- Configuring Module: cudnn --------
    creating /tmp/tmpwnclorrz/tmp
    creating /tmp/tmpwnclorrz/tmp/tmpwnclorrz
    gcc -pthread -B /home/leofang/miniconda3/envs/cupy_dev_cuda11/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -D_FORCE_INLINES=1 -I/usr/local/cuda-11.0/include/cub -I/home/leofang/cupy_cuda11/install/../cupy/core/include -I/usr/local/cuda-11.0/include -c /tmp/tmpwnclorrz/a.cpp -o /tmp/tmpwnclorrz/tmp/tmpwnclorrz/a.o
    cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
    In file included from /usr/local/cuda-11.0/include/channel_descriptor.h:61:0,
                     from /usr/local/cuda-11.0/include/cuda_runtime.h:95,
                     from /usr/include/cudnn.h:64,
                     from /tmp/tmpwnclorrz/a.cpp:1:
    /usr/local/cuda-11.0/include/cuda_runtime_api.h:2306:36: error: use of enum ‘cudaStreamAttrID’ without previous declaration
             cudaStream_t hStream, enum cudaStreamAttrID attr,
                                        ^~~~~~~~~~~~~~~~
    /usr/local/cuda-11.0/include/cuda_runtime_api.h:2330:36: error: use of enum ‘cudaStreamAttrID’ without previous declaration
             cudaStream_t hStream, enum cudaStreamAttrID attr,
                                        ^~~~~~~~~~~~~~~~
    /usr/local/cuda-11.0/include/cuda_runtime_api.h:2631:88: error: use of enum ‘cudaStreamCaptureMode’ without previous declaration
     extern __host__ cudaError_t CUDARTAPI cudaStreamBeginCapture(cudaStream_t stream, enum cudaStreamCaptureMode mode);
                                                                                            ^~~~~~~~~~~~~~~~~~~~~
    /usr/local/cuda-11.0/include/cuda_runtime_api.h:2682:80: error: use of enum ‘cudaStreamCaptureMode’ without previous declaration
     extern __host__ cudaError_t CUDARTAPI cudaThreadExchangeStreamCaptureMode(enum cudaStreamCaptureMode *mode);
                                                                                    ^~~~~~~~~~~~~~~~~~~~~
    In file included from /usr/local/cuda-11.0/include/channel_descriptor.h:61:0,
                     from /usr/local/cuda-11.0/include/cuda_runtime.h:95,
                     from /usr/include/cudnn.h:64,
                     from /tmp/tmpwnclorrz/a.cpp:1:
    /usr/local/cuda-11.0/include/cuda_runtime_api.h:8758:10: error: use of enum ‘cudaKernelNodeAttrID’ without previous declaration
         enum cudaKernelNodeAttrID attr,
              ^~~~~~~~~~~~~~~~~~~~
    /usr/local/cuda-11.0/include/cuda_runtime_api.h:8782:10: error: use of enum ‘cudaKernelNodeAttrID’ without previous declaration
         enum cudaKernelNodeAttrID attr,
              ^~~~~~~~~~~~~~~~~~~~
    /usr/local/cuda-11.0/include/cuda_runtime_api.h:9740:145: error: use of enum ‘cudaGraphExecUpdateResult’ without previous declaration
     extern __host__ cudaError_t CUDARTAPI cudaGraphExecUpdate(cudaGraphExec_t hGraphExec, cudaGraph_t hGraph, cudaGraphNode_t *hErrorNode_out, enum cudaGraphExecUpdateResult *updateResult_out);
                                                                                                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~
    /usr/local/cuda-11.0/include/cuda_runtime_api.h:9987:49: error: ‘cudaFunction_t’ was not declared in this scope
     extern __host__ cudaError_t cudaGetFuncBySymbol(cudaFunction_t* functionPtr, const void* symbolPtr);
                                                     ^~~~~~~~~~~~~~
    /usr/local/cuda-11.0/include/cuda_runtime_api.h:9987:49: note: suggested alternative: ‘cudaFuncCache’
     extern __host__ cudaError_t cudaGetFuncBySymbol(cudaFunction_t* functionPtr, const void* symbolPtr);
                                                     ^~~~~~~~~~~~~~
                                                     cudaFuncCache
    /usr/local/cuda-11.0/include/cuda_runtime_api.h:9987:65: error: ‘functionPtr’ was not declared in this scope
     extern __host__ cudaError_t cudaGetFuncBySymbol(cudaFunction_t* functionPtr, const void* symbolPtr);
                                                                     ^~~~~~~~~~~
    /usr/local/cuda-11.0/include/cuda_runtime_api.h:9987:78: error: expected primary-expression before ‘const’
     extern __host__ cudaError_t cudaGetFuncBySymbol(cudaFunction_t* functionPtr, const void* symbolPtr);
                                                                                  ^~~~~
    /usr/local/cuda-11.0/include/cuda_runtime_api.h:9987:99: error: expression list treated as compound expression in initializer [-fpermissive]
     extern __host__ cudaError_t cudaGetFuncBySymbol(cudaFunction_t* functionPtr, const void* symbolPtr);
                                                                                                       ^
    command 'gcc' failed with exit status 1
```

--- a/tests/cupy_tests/test_cudnn.py
+++ b/tests/cupy_tests/test_cudnn.py
@@ -255,6 +255,12 @@ class TestConvolutionBackwardFilter(unittest.TestCase):
                 (ndim > 2 and version < 6000) or
                 (ndim > 2 and self.dtype == numpy.float64)):
             self.err = libcudnn.CuDNNError
+        elif (8000 <= version and
+              self.max_workspace_size == 0 and
+              int(cupy.cuda.device.get_compute_capability()) < 70 and
+              self.groups > 1 and ndim > 2 and
+              self.dtype == numpy.float16):
+            self.err = RuntimeError
         self._workspace_size = cudnn.get_max_workspace_size()
         cudnn.set_max_workspace_size(self.max_workspace_size)
 
@@ -331,6 +337,11 @@ class TestConvolutionBackwardData(unittest.TestCase):
                 (ndim > 2 and version < 6000) or
                 (ndim > 2 and self.dtype == numpy.float64)):
             self.err = libcudnn.CuDNNError
+        elif (8000 <= version and
+              int(cupy.cuda.device.get_compute_capability()) < 70 and
+              self.dilate > 1 and self.groups > 1 and ndim > 2 and
+              self.dtype == numpy.float16):
+            self.err = RuntimeError
         self._workspace_size = cudnn.get_max_workspace_size()
         cudnn.set_max_workspace_size(self.max_workspace_size)
 

